compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -29,7 +29,13 @@ crate fn mir_built<'tcx>(
         return tcx.mir_built(def);
     }
 
-    tcx.alloc_steal_mir(mir_build(tcx, def))
+    let mut body = mir_build(tcx, def);
+    if def.const_param_did.is_some() {
+        assert!(matches!(body.source.instance, ty::InstanceDef::Item(_)));
+        body.source = MirSource::from_instance(ty::InstanceDef::Item(def.to_global()));
+    }
+
+    tcx.alloc_steal_mir(body)
```

Comment:
@lcnr what should this look like in your eyes? It seems like we should do this when the MIR is built instead, but that means calling `try_upgrade` outside of a query?

Reply:
```rust
let mut body = mir_built(txc, def);
```

Reply:
we already have `def` as an input to `mir_built`, so we should be able to use the correct `source` right from the start here.

Reply:
That causes cycle errors during type-checking.

Reply:
why do you think that we should call `try_upgrade` here?

Reply:
Define "here"? There's already a call to `try_upgrade` above.

Reply:
This commit is a hack that I stumbled on after some trial and error to avoid cycle errors on some const-generics tests. I was hoping to learn what the correct approach would look like.

Reply:
>  I meant inside rustc_mir_builder::Builder::build() or so.

yeah, we should never have to call `try_upgrade` inside of `mir_built` after the initial call at the start of the query. If you can get the failing state again I would gladly look into what's going wrong there

Reply:
```
error[E0391]: cycle detected when type-checking `main`
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:33
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                 ^^^^^^^^^^^^^
   |
note: ...which requires simplifying constant for the type system `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
note: ...which requires const-evaluating + checking `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
   = note: ...which requires normalizing `main::{constant#9}::promoted[0]`...
note: ...which requires resolving instance `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing the optional const parameter of `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
   = note: ...which again requires type-checking `main`, completing the cycle
   = note: cycle used when type-checking all item bodies

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0391.
For more information about an error, try `rustc --explain E0308`.

------------------------------------------



failures:
    [ui] ui/const-generics/slice-const-param-mismatch.rs#full
    [ui] ui/const-generics/slice-const-param.rs#full
```

Cycle along with the two failing tests. This is what happens with *just* the first commit.

Reply:
```
   = note: ...which requires normalizing `main::{constant#9}::promoted[0]`...
```
Is probably where this goes wrong, so we somehow are not  using the full `def` for promoteds.

Not at home rn, will look at where exactly this goes wrong tomorrow.

Reply:
So your solution is probably the easiest for now...

T the actual mir builder currently only uses a `DefId` and has quite a few entry points.
After we merge this I am going to look into a good way to start with the correct `WithOptConstParam` in the `build::Builder`.
The way I expect to do this is a bit like the following but without using `WithOptConstParam::unknown` there, which is a bit more invasive than I would like in this PR: https://github.com/lcnr/rust/commit/90615386cc3d8a3a7f3edf7bd0428dcc3f987bd3

Reply:
Apart from this I think this is mostly ready for merge :thinking: is there still something you want to do in that PR?